### PR TITLE
fix: pool params query

### PR DESCRIPTION
**Pull request by nghuyenthevinh2000**
Closes: #1089 

The code here can solve issues addressed at: https://github.com/osmosis-labs/osmosis/issues/1089

Here is a picture to prove it

Screenshot from 2022-04-22 01-20-06

## What is the purpose of the change

*Add a description of the overall background and high level changes that this PR introduces *

*(E.g.: This pull request improves documation of area A by adding ....*

## Brief change log

*(for example:)*
 
  - *The metadata is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *Daemons retrieve the RPC data from the blob cache*

## Testing and Verifying

*(Please pick one of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added unit test that validates ...*
  - *Added integration tests for end-to-end deployment with ...*
  - *Extended integration test for ...*
  - *Manually verified the change by ...*

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/gamm/pool-models/balancer/codec.go [hunk 1]
```diff
@@ -26,6 +26,7 @@ func RegisterInterfaces(registry codectypes.InterfaceRegistry) {
 	registry.RegisterImplementations(
 		(*sdk.Msg)(nil),
 		&MsgCreateBalancerPool{},
+		&PoolParams{},
```
Comment (ValarDragon) [inline x/gamm/pool-models/balancer/codec.go:29]: PoolParams doesnt implement the message interface though?

File: x/gamm/client/cli/query.go [hunk 2]
```diff
@@ -13,6 +13,8 @@ import (
 	"gopkg.in/yaml.v2"
 
 	"github.com/osmosis-labs/osmosis/v7/x/gamm/types"
+
```
Comment (alexanderbez) [inline x/gamm/client/cli/query.go:16]:
```suggestion
```

File: x/gamm/pool-models/balancer/msgs.go [hunk 3]
```diff
@@ -98,3 +99,22 @@ func (msg MsgCreateBalancerPool) CreatePool(ctx sdk.Context, poolID uint64) (typ
 	poolI, err := NewBalancerPool(poolID, *msg.PoolParams, msg.PoolAssets, msg.FuturePoolGovernor, ctx.BlockTime())
 	return &poolI, err
 }
+
+var _ sdk.Msg = &PoolParams{}
```
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/msgs.go:103]: Why are we making this change?

**Comment by ValarDragon** [comment]
The approach of this PR is making pool params satisfy the message interface. What should likely instead be done is just registering this struct onto the proto codec directly.

File: x/gamm/pool-models/balancer/codec.go (hunk 1 above)
Comment (nghuyenthevinh2000) [reply x/gamm/pool-models/balancer/codec.go:33]: yes, I have added PoolParams directly to codec like your suggestion

Comment (mattverse) [reply x/gamm/pool-models/balancer/codec.go:33]: I think what @ValarDragon was saying was that `PoolParams` does not follow implementations of proto.Message. 

Structs should be implemented upon specific interfaces via `RegisterImplementations`, @ValarDragon  do you think we should make an interface for pool params?

**Comment by ValarDragon** [comment]
So the problem being faced here, is that the PoolParams is getting serialized and then deserialized as a types.Any: https://github.com/osmosis-labs/osmosis/blob/6f50a357b2f0c441f64e61cd2f4da9066e1af1e0/x/gamm/types/query.pb.go#L351-L353

This means that we need to register it as implementing any interface. I think it suffices to say the struct implements `interface{}` rather than sdk.Msg. Can you try that and see if it fixes the issue?

Also btw, `coverage.txt` needs to be deleted

**Comment by mattverse** [comment]
I think a better approach is creating a pool params interface rather than using a emtpy interface wdyt?

**Comment by ValarDragon** [comment]
Why would we have an interface? The problem here is the codec doesn't know about this route, and it can learn about the route by registering it as an implementation of any interface, such as `interface{}`

**Comment by nghuyenthevinh2000** [comment]
wouldn't it be better to use `proto.Message` interface rather than just `interface{}` cause `PoolParams{}` already satisfied `proto.Message` by default when it was first generated.

`registry.RegisterImplementations(
		(*proto.Message)(nil),
		&PoolParams{},
	)
`

File: x/gamm/client/cli/query.go [hunk 4]
```diff
@@ -216,7 +217,10 @@ $ %s query gamm pool-params 1
 			}
 
 			if clientCtx.OutputFormat == "text" {
-				out, err := yaml.Marshal(res.GetParams())
+				poolParams := &balancer.PoolParams{}
+				poolParams.Unmarshal(res.GetParams().Value)
```
Comment (alexanderbez) [inline x/gamm/client/cli/query.go:221]: we need to check the error here.

**Comment by ValarDragon** [comment]
oh good point, using `proto.Message` as the interface makes sense as well.

**Review by mattverse — APPROVED** [review]
Awesome! LGTM!

**Comment by p0mvn** [comment]
Hi @nghuyenthevinh2000 . Would you mind updating the PR name to something like "fix: pool params query", please? I'm requesting this so that it is easier to follow in Github history, once merged. Thank you!

**Comment by nghuyenthevinh2000** [comment]
I have renamed as requested

File: x/gamm/client/cli/query.go [hunk 5]
```diff
@@ -216,7 +217,13 @@ $ %s query gamm pool-params 1
 			}
 
 			if clientCtx.OutputFormat == "text" {
-				out, err := yaml.Marshal(res.GetParams())
+				poolParams := &balancer.PoolParams{}
+				err := poolParams.Unmarshal(res.GetParams().Value)
+				if err != nil {
```
Comment (p0mvn) [inline x/gamm/client/cli/query.go:222]:
```suggestion
				if err := poolParams.Unmarshal(res.GetParams().Value); err != nil {
```

**Review by p0mvn — APPROVED** [review]

**Pull request merged**